clab/config/functions.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil || !matched {
+		return "", fmt.Errorf("value %s does not match regex %s %v", vals, format, err)
+	}
+
+	return "", nil
+}
+
+var funcMap = map[string]interface{}{
+	"optional": func(val interface{}, format string) (interface{}, error) {
+		if val == nil {
+			return "", nil
+		}
+		return expectFunc(val, format)
+	},
+	"expect": expectFunc,
+	// "require": func(val interface{}) (interface{}, error) {
+	// 	if val == nil {
+	// 		return nil, errors.New("required value not set")
+	// 	}
+	// 	return val, nil
+	// },
+	"ip": func(val interface{}) (interface{}, error) {
+		s := fmt.Sprintf("%v", val)
+		a := strings.Split(s, "/")
+		return a[0], nil
+	},
+	"ipmask": func(val interface{}) (interface{}, error) {
+		s := fmt.Sprintf("%v", val)
+		a := strings.Split(s, "/")
+		return a[1], nil
+	},
+	"default": func(in ...interface{}) (interface{}, error) {
+		if len(in) < 2 {
+			return nil, fmt.Errorf("default value expected")
+		}
+		if len(in) > 2 {
+			return nil, fmt.Errorf("too many arguments")
+		}
+
+		val := in[len(in)-1]
+		def := in[0]
+
+		switch v := val.(type) {
+		case nil:
+			return def, nil
+		case string:
+			if v == "" {
+				return def, nil
+			}
+		case bool:
+			if !v {
+				return def, nil
+			}
+		}
+		// if val == nil {
+		// 	return def, nil
+		// }
+
+		// If we have a input value, do some type checking
+		tval, tdef := typeof(val), typeof(def)
+		if tval == "string" && tdef == "int" {
+			if _, err := strconv.Atoi(val.(string)); err == nil {
+				tval = "int"
+			}
+			if tdef == "str" {
+				if _, err := strconv.Atoi(def.(string)); err == nil {
+					tdef = "int"
+				}
+			}
+		}
+		if tdef != tval {
+			return val, fmt.Errorf("expected type %v, got %v (value=%v)", tdef, tval, val)
+		}
+
+		// Return the value
+		return val, nil
+	},
+	"contains": func(substr string, str string) (interface{}, error) {
+		return strings.Contains(fmt.Sprintf("%v", str), fmt.Sprintf("%v", substr)), nil
+	},
+	"split": func(sep string, val interface{}) (interface{}, error) {
+		// Start and end values
+		if val == nil {
+			return []interface{}{}, nil
+		}
+		if sep == "" {
+			sep = " "
+		}
+
+		v := fmt.Sprintf("%v", val)
+
+		res := strings.Split(v, sep)
+		r := make([]interface{}, len(res))
+		for i, p := range res {
+			r[i] = p
+		}
+		return r, nil
+	},
+	"join": func(sep string, val interface{}) (interface{}, error) {
+		if sep == "" {
+			sep = " "
+		}
+		// Start and end values
+		switch v := val.(type) {
+		case []interface{}:
+			if val == nil {
+				return "", nil
+			}
+			res := make([]string, len(v))
+			for i, v := range v {
+				res[i] = fmt.Sprintf("%v", v)
+			}
+			return strings.Join(res, sep), nil
+		case []string:
+			return strings.Join(v, sep), nil
+		case []int, []int16, []int32:
+			return strings.Trim(strings.ReplaceAll(fmt.Sprint(v), " ", sep), "[]"), nil
+		}
+		return nil, fmt.Errorf("expected array [], got %v", val)
+	},
+	"slice": func(start, end int, val interface{}) (interface{}, error) {
```

Comment:
since template package has a `slice` builtin function, maybe you need to rename this one to something else?
https://pkg.go.dev/text/template#hdr-Functions

Reply:
They are there, but their arguments all seem wrong, especially if you want to chain them:

> A pipeline may be "chained" by separating a sequence of commands with pipeline characters '|'. In a chained pipeline, the result of each command is passed as the last argument of the following command. The output of the final command in the pipeline is the value of the pipeline.

So my version of the functions all accept them as the LAST argument, the built-in all accept them as the FIRST argument

One of my test cases: `{{ "1.1.1.1/32" | split "/" | slice 1 2 | join "" }}` --> `32` [here](https://github.com/kellerza/containerlab/blob/config/clab/config/functions_test.go#L40)

Now try something similar in go template [here](https://play.golang.org/p/U54aSLXhjs8). There is a line you can uncomment to make it work...

Reply:
thanks, that is clear, and I don't object having custom funcs like that. What I am  proposing though is to avoid shadowing the embedded function names like `slice` and `index` and instead use some custom name

Users are in charge of writing the templates, and some of them will rely on `index` to behave the same way it behaves in stdlib, so overwriting this with custom functionality may be dangerous and frustrating.

What if we just give those custom func some distinctive names like `index-pipe`, `slice-pipe`?

Reply:
or maybe less verbose `_index` and `_slice` that will keep the original meaning but will not clash with the built-in functions